cty/function/stdlib/collection.go

```diff
@@ -953,7 +953,7 @@ var SetProductFunc = function.New(&function.Spec{
 			// If any of the arguments was an empty collection then our result
 			// is also an empty collection, which we'll short-circuit here.
 			if retType.IsListType() {
-				return cty.ListValEmpty(ety).Mark(retMarks), nil
+				return cty.ListValEmpty(ety), nil
 			}
 			return cty.SetValEmpty(ety).Mark(retMarks), nil
```

Comment:
This is not the correct change here, as it will result in dropping marks from input values.

Instead, I think the crash is caused by calling `Mark` with a `cty.ValueMarks` value, instead of calling `WithMarks`. Unfortunately the compiler didn't catch this because `Mark` accepts `interface{}`.

Would you mind fixing this up to call `WithMarks` for both the empty list and empty set branches? Adding tests cases to cover this would be great too. Input of two empty lists marked "a" and "b" respectively should result in an empty list marked with both "a" and "b", and the same for sets.